app/models/query.rb

```diff
@@ -208,6 +209,32 @@ def validate_work_package_filters
     end
   end
 
+  def validate_columns
+    available_names = available_columns.map(&:name).map(&:to_s)
+
+    column_names.each do |name|
+      unless available_names.include? name.to_s
+        errors.add :column_names, I18n.t(:error_invalid_query_column, value: name)
+      end
+    end
+  end
+
+  def validate_sort_criteria
+    available_criteria = sortable_columns.map(&:name).map(&:to_s)
+
+    sort_criteria.each do |name, _dir|
+      unless available_criteria.include? name.to_s
+        errors.add :sort_criteria, I18n.t(:error_invalid_sort_criterion, value: name)
+      end
+    end
+  end
+
+  def validate_group_by
+    unless group_by.nil? || groupable_columns.map(&:name).map(&:to_s).include?(group_by.to_s)
```

Comment:
group_by seems to be `""` when not grouping and coming from the DB. Thus we would need to check for `blank?` instead.